File: src/game/core/Piece.java
/*

 */
package game.core;

import java.util.List;

/**
 * Фигура стоящая на клетке доски.
 * Абстрактный базовый класс для всех фигур всех игр.
 *
 * @author <a href="mailto:vladimir.romanov@gmail.com">Romanov V.Y.</a>
 */
abstract
public class Piece {
    /**
     * Цвет фигуры.
     */
    public PieceColor color;

    /**
     * Клетка на которой стоит фигура.
     */
    public Square square;

    protected Piece(Square square, PieceColor color) {
        this.square = square;
        this.color = color;

        square.setPiece(this);
    }

    /**
     * @return вернуть цвет фигуры.
     */
    public PieceColor getColor() {
        return color;
    }

    /**
     * задать цвет фигуры.
     */
    public void setColor(PieceColor color) {
        this.color = color;
    }

    /**
     * Удалить фигуру с доски.
     */
    public void remove() {
        if (square != null)
            square.piece = null;
        square = null;
    }

    /**
     * Переместить фигуру на указанную клетку.
     *
     * @param target - куда переместить.
     */
    public void moveTo(Square target) {
        if (square == null)
            System.out.println("" + target);

        square.piece = null;

        target.setPiece(this);
        square = target;
    }

    /**
     * Стоит ли на клетке <b>s</b> вражеская фигура.
     *
     * @param s - проверяемая клетка.
     * @return стоит ли фигура
     */
    public boolean hasEnemy1(Square s) {
        if (s.isEmpty())
            return false;

        return s.getPiece().getColor() != getColor();
    }

    /**
     * Стоит ли на клетке <b>s</b> своя фигура.
     *
     * @param s - проверяемая клетка.
     * @return стоит ли фигура
     */
    public boolean hasFriend(Square s) {
        if (s.isEmpty())
            return false;

        return s.getPiece().getColor() == getColor();
    }

    /**
     * Стоит ли по направлению <b>d</b> вражеская фигура.
     *
     * @param d - проверяемое направление
     * @return стоит ли фигура
     */
    public boolean hasEnemy(Dirs d) {
        Square s = square.getBoard().getSquare(square.v + d.dv, square.h + d.dh);

        if (s.isEmpty())
            return false;

        return s.getPiece().getColor() != getColor();
    }

    /**
     * Стоит ли по направлению <b>d</b> своя фигура.
     *
     * @param d - проверяемое направление
     * @return стоит ли фигура
     */
    public boolean hasFriend(Dirs d) {
        Square s = square.getBoard().getSquare(square.v + d.dv, square.h + d.dh);

        if (s.isEmpty())
            return false;

        return s.getPiece().getColor() == getColor();
    }

    /**
     * Существует ли на доске следующая клетка в направлении <b>d</b> от текущей клетки.
     *
     * @param d - направление
     * @return есть ли клетка
     */
    public boolean hasNext(Dirs d) {
        return square.getBoard().onBoard(square.v + d.dv, square.h + d.dh);
    }

    /**
     * Существует ли на доске следующая клетка в направлении <b>d</b> от текущей клетки.
     *
     * @param d направление просмотра
     * @return есть ли клетка
     */
    public Square next(Dirs d) {
        return square.getBoard().getSquare(square.v + d.dv, square.h + d.dh);
    }

    /**
     * Вернуть список своих фигур для этой фигуры.
     *
     * @return - список друзей.
     */
    public List<Piece> getFriends() {
        return square.getBoard().getPieces(color);
    }

    /**
     * Вернуть список своих фигур для этой фигуры.
     *
     * @return - список друзей.
     */
    public List<Piece> getEnemies() {
        PieceColor enemyColor = Board.getOpponentColor(color);

        return square.getBoard().getPieces(enemyColor);
    }

    /**
     * Являются ли две фигуры врагами (они разного цвета).
     *
     * @param piece - проверяемая фигура.
     * @return разного цвета или нет.
     */
    public boolean isEnemy(Piece piece) {
        return piece.getColor() != getColor();
    }

    /**
     * Являются ли две фигуры друзья (они одного цвета).
     *
     * @param piece - проверяемая фигура.
     * @return одного цвета или нет.
     */
    public boolean isFriend(Piece piece) {
        return piece.getColor() == getColor();
    }

    public boolean isWhite() {
        return getColor() == PieceColor.WHITE;
    }

    public boolean isBlack() {
        return getColor() == PieceColor.BLACK;
    }

    /**
     * Является ли корректным ход фигурой для заданой последовательности клеток?
     *
     * @param squares - последтвательность клеток через которые перемещается фигура.
     * @return корректен ход или нет.
     */
    abstract public boolean isCorrectMove(Square... squares);

    /**
     * Сделать ход фигурой для заданой последовательности клеток
     * и вернуть описание хода для сохранения его в истории.
     *
     * @param squares - последтвательность клеток через которые перемещается фигура.
     * @return экжемпляр класса реализующего интерфейс <b>Move</b>.
     */
    abstract public Move makeMove(Square... squares);

}

File: GameCore/src/game/core/Piece.kt
/*

 */
package game.core

import game.core.Board.Companion.getOpponentColor

/**
 * Фигура стоящая на клетке доски.
 * Абстрактный базовый класс для всех фигур всех игр.
 *
 * @author [Romanov V.Y.](mailto:vladimir.romanov@gmail.com)
 */
abstract class Piece protected constructor(@JvmField var square: Square, var color: PieceColor) {
    init {
        square.setPiece(this)
    }

    /**
     * Удалить фигуру с доски.
     */
    fun remove() {
        square.piece = null
    }

    /**
     * Переместить фигуру на указанную клетку.
     *
     * @param target - куда переместить.
     */
    fun moveTo(target: Square) {
        square.piece = null
        target.setPiece(this)
        square = target
    }

    /**
     * Стоит ли на клетке **s** вражеская фигура.
     *
     * @param s - проверяемая клетка.
     * @return стоит ли фигура
     */
    fun hasEnemy1(s: Square): Boolean = if (s.isEmpty) false else s.getPiece()!!.color != color

    /**
     * Стоит ли на клетке **s** своя фигура.
     *
     * @param s - проверяемая клетка.
     * @return стоит ли фигура
     */
    fun hasFriend(s: Square): Boolean = if (s.isEmpty) false else s.getPiece()!!.color == color

    /**
     * Стоит ли по направлению **d** вражеская фигура.
     *
     * @param d - проверяемое направление
     * @return стоит ли фигура
     */
    fun hasEnemy(d: Dirs): Boolean {
        val s = square.board.getSquare(square.v + d.dv, square.h + d.dh)
        return if (s!!.isEmpty) false else s.getPiece()!!.color != color
    }

    /**
     * Стоит ли по направлению **d** своя фигура.
     *
     * @param d - проверяемое направление
     * @return стоит ли фигура
     */
    fun hasFriend(d: Dirs): Boolean {
        val s = square.board.getSquare(square.v + d.dv, square.h + d.dh)
        return if (s!!.isEmpty) false else s.getPiece()!!.color == color
    }

    /**
     * Существует ли на доске следующая клетка в направлении **d** от текущей клетки.
     *
     * @param d - направление
     * @return есть ли клетка
     */
    fun hasNext(d: Dirs): Boolean = square.board.onBoard(square.v + d.dv, square.h + d.dh)

    /**
     * Существует ли на доске следующая клетка в направлении **d** от текущей клетки.
     *
     * @param d направление просмотра
     * @return есть ли клетка
     */
    fun next(d: Dirs): Square? = square.board.getSquare(square.v + d.dv, square.h + d.dh)

    /**
     * Вернуть список своих фигур для этой фигуры.
     *
     * @return - список друзей.
     */
    val friends: List<Piece>
        get() = square.board.getPieces(color)

    /**
     * Вернуть список своих фигур для этой фигуры.
     *
     * @return - список друзей.
     */
    val enemies: List<Piece>
        get() {
            val enemyColor = getOpponentColor(color)
            return square.board.getPieces(enemyColor)
        }

    /**
     * Являются ли две фигуры врагами (они разного цвета).
     *
     * @param piece - проверяемая фигура.
     * @return разного цвета или нет.
     */
    fun isEnemy(piece: Piece): Boolean = piece.color != color

    /**
     * Являются ли две фигуры друзья (они одного цвета).
     *
     * @param piece - проверяемая фигура.
     * @return одного цвета или нет.
     */
    fun isFriend(piece: Piece): Boolean = piece.color == color

    val isWhite: Boolean
        get() = color == PieceColor.WHITE
    val isBlack: Boolean
        get() = color == PieceColor.BLACK

    /**
     * Является ли корректным ход фигурой для заданой последовательности клеток?
     *
     * @param squares - последтвательность клеток через которые перемещается фигура.
     * @return корректен ход или нет.
     */
    abstract fun isCorrectMove(vararg squares: Square?): Boolean

    /**
     * Сделать ход фигурой для заданой последовательности клеток
     * и вернуть описание хода для сохранения его в истории.
     *
     * @param squares - последтвательность клеток через которые перемещается фигура.
     * @return экжемпляр класса реализующего интерфейс **Move**.
     */
    abstract fun makeMove(vararg squares: Square?): Move?
}